### PR TITLE
[python-package] fix some warnings from mypy

### PR DESCRIPTION
**Pull request by tara-jawahar**
Fixed some mypy type warnings in setup.py, dask.py, and sklearn.py

### CODE REVIEW

File: python-package/setup.py [hunk 1]
```diff
@@ -322,7 +322,7 @@ def run(self):
     if os.path.isfile(os.path.join(CURRENT_DIR, os.path.pardir, 'VERSION.txt')):
         copy_file(os.path.join(CURRENT_DIR, os.path.pardir, 'VERSION.txt'),
                   os.path.join(CURRENT_DIR, 'lightgbm', 'VERSION.txt'),
-                  verbose=0)
+                  verbose=False)
```
Comment (jameslamb) [inline python-package/setup.py:325]: Was this to fix this warning?

> python-package/setup.py:325: error: Argument "verbose" to "copy_file" has incompatible type "int"; expected "bool"

I think this is another case where `mypy` is getting confused. I can see in the definition of `copy_files()` that verbose is an int: https://github.com/python/cpython/blob/9eb11a139fac5514d8456626806a68b3e3b7eafb/Lib/distutils/file_util.py#L68.

So can you please add a `# type: ignore` comment here to tell `mypy` to trust us on this line?

```suggestion
                  verbose=0)  # type:ignore
```

File: python-package/lightgbm/dask.py [hunk 2]
```diff
@@ -27,7 +27,7 @@
 _PredictionDtype = Union[Type[np.float32], Type[np.float64], Type[np.int32], Type[np.int64]]
 
 
-def _find_open_port(worker_ip: str, local_listen_port: int, ports_to_skip: Iterable[int]) -> int:
+def _find_open_port(worker_ip: str, local_listen_port: int, ports_to_skip: Iterable[int]) -> Optional[int]:
```
Comment (jameslamb) [inline python-package/lightgbm/dask.py:30]: Is this to fix this warning?

> python-package/lightgbm/dask.py:71: error: Incompatible return value type (got "Optional[int]", expected "int")

I don't think this is the right fix. I think that `mypy` is (rightly) telling us that I wrote `_find_open_port()` in a way that makes it a little confusing to reason about the return type of `out_port`. Can you please change this back to `int`, and remove `out_port = None` on line 54? I think that should fix that warning.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks very much for taking the time to contribute!

I left a few suggestions, most of which can be accepted directly in the browser.

After your next commit, can you please add a comment on this pull request with the output of running `mypy --ignore-missing-imports python-package/?`.

Also I hope you don't mind, but I've changed the title of this pull request to something that will be easy to understand in a changelog. We automatically generate changelogs from PR titles in this project: https://github.com/microsoft/LightGBM/releases/tag/untagged-e810435e8b97e053e761

**Comment by tara-jawahar** [comment]
Output of mypy --ignore-missing-imports python-package/ :

python-package/lightgbm/compat.py:14: error: Name 'pd_Series' already defined (possibly by an import)
python-package/lightgbm/compat.py:19: error: Name 'pd_DataFrame' already defined (possibly by an import)
python-package/lightgbm/compat.py:52: error: Name 'dt_DataTable' already defined on line 45
python-package/lightgbm/compat.py:128: error: Name 'dask_Array' already defined (possibly by an import)
python-package/lightgbm/compat.py:133: error: Name 'dask_DataFrame' already defined (possibly by an import)
python-package/lightgbm/compat.py:138: error: Name 'dask_Series' already defined (possibly by an import)
python-package/lightgbm/callback.py:28: error: First argument to namedtuple() should be 'CallbackEnv', not 'LightGBMCallbackEnv'
python-package/lightgbm/sklearn.py:795: error: Value of type "Optional[str]" is not indexable
python-package/lightgbm/sklearn.py:795: error: Item "None" of "Optional[str]" has no attribute "find"
python-package/lightgbm/sklearn.py:796: error: Value of type "Optional[str]" is not indexable
python-package/lightgbm/sklearn.py:796: error: Item "None" of "Optional[str]" has no attribute "find"
python-package/lightgbm/sklearn.py:868: error: Value of type "Optional[str]" is not indexable
python-package/lightgbm/sklearn.py:868: error: Item "None" of "Optional[str]" has no attribute "find"
python-package/lightgbm/sklearn.py:869: error: Value of type "Optional[str]" is not indexable
python-package/lightgbm/sklearn.py:869: error: Item "None" of "Optional[str]" has no attribute "find"
python-package/lightgbm/sklearn.py:991: error: Value of type "Optional[str]" is not indexable
python-package/lightgbm/sklearn.py:991: error: Item "None" of "Optional[str]" has no attribute "find"
python-package/lightgbm/sklearn.py:992: error: Value of type "Optional[str]" is not indexable
python-package/lightgbm/sklearn.py:992: error: Item "None" of "Optional[str]" has no attribute "find"
python-package/lightgbm/dask.py:284: error: Incompatible return value type (got "Dict[str, Any]", expected "LGBMModel")
python-package/lightgbm/dask.py:452: error: "_DaskLGBMModel" has no attribute "get_params"
python-package/lightgbm/dask.py:465: error: "_DaskLGBMModel" has no attribute "set_params"
python-package/lightgbm/dask.py:471: error: "_DaskLGBMModel" has no attribute "get_params"
python-package/lightgbm/dask.py:477: error: Item "_DaskLGBMModel" of "Union[_DaskLGBMModel, LGBMModel]" has no attribute "get_params"
python-package/lightgbm/dask.py:487: error: Signature of "fit" incompatible with supertype "LGBMClassifier"
python-package/lightgbm/dask.py:487: error: Signature of "fit" incompatible with supertype "LGBMModel"
python-package/lightgbm/dask.py:496: error: Incompatible return value type (got "_DaskLGBMModel", expected "DaskLGBMClassifier")
python-package/lightgbm/dask.py:512: error: Signature of "predict" incompatible with supertype "LGBMClassifier"
python-package/lightgbm/dask.py:512: error: Signature of "predict" incompatible with supertype "LGBMModel"
python-package/lightgbm/dask.py:523: error: Signature of "predict_proba" incompatible with supertype "LGBMClassifier"
python-package/lightgbm/dask.py:548: error: Signature of "fit" incompatible with supertype "LGBMRegressor"
python-package/lightgbm/dask.py:548: error: Signature of "fit" incompatible with supertype "LGBMModel"
python-package/lightgbm/dask.py:557: error: Incompatible return value type (got "_DaskLGBMModel", expected "DaskLGBMRegressor")
python-package/lightgbm/dask.py:573: error: Signature of "predict" incompatible with supertype "LGBMModel"
python-package/lightgbm/dask.py:597: error: Signature of "fit" incompatible with supertype "LGBMRanker"
python-package/lightgbm/dask.py:597: error: Signature of "fit" incompatible with supertype "LGBMModel"
python-package/lightgbm/dask.py:611: error: Incompatible return value type (got "_DaskLGBMModel", expected "DaskLGBMRanker")
python-package/lightgbm/dask.py:622: error: Item "None" of "Optional[str]" has no attribute "partition"
python-package/lightgbm/dask.py:628: error: Signature of "predict" incompatible with supertype "LGBMModel"
Found 39 errors in 4 files (checked 10 source files)

File: python-package/lightgbm/sklearn.py [hunk 3]
```diff
@@ -792,12 +792,13 @@ def fit(self, X, y,
...
+    assert isinstance(_base_doc, str)
     _base_doc = (_base_doc[:_base_doc.find('group :')]
-                 + _base_doc[_base_doc.find('eval_set :'):])
+                + _base_doc[_base_doc.find('eval_set :'):])
     _base_doc = (_base_doc[:_base_doc.find('eval_class_weight :')]
-                 + _base_doc[_base_doc.find('eval_init_score :'):])
+                + _base_doc[_base_doc.find('eval_init_score :'):])
     fit.__doc__ = (_base_doc[:_base_doc.find('eval_group :')]
-                   + _base_doc[_base_doc.find('eval_metric :'):])
+                + _base_doc[_base_doc.find('eval_metric :'):])
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply python-package/lightgbm/sklearn.py:800]: I saw that this was resolved but I'm still curious. What is the reason for these changes? If they aren't related to `mypy`, can you please revert them?

File: python-package/lightgbm/dask.py [hunk 4]
```diff
@@ -564,7 +564,7 @@ def fit(
         )
 
     _base_doc = LGBMRegressor.fit.__doc__
-    _before_init_score, _init_score, _after_init_score = _base_doc.partition('init_score :')
+    _before_init_score, _init_score, _after_init_score = _base_doc.partition('init_score :') # type: ignore
```
Comment (jameslamb) [inline python-package/lightgbm/dask.py:567]: ah, do these calls fix this?

> python-package/lightgbm/dask.py:567: error: Item "None" of "Optional[str]" has no attribute "partition"

File: python-package/lightgbm/dask.py [hunk 5]
```diff
@@ -280,11 +280,11 @@ def _train(
     wait(parts)
 
     for part in parts:
-        if part.status == 'error':
+        if part.status == 'error': # type: ignore
```
Comment (jameslamb) [inline python-package/lightgbm/dask.py:283]: oh nice! Does this fix  this?

> python-package/lightgbm/dask.py:283: error: "Dict[str, Any]" has no attribute "status"

File: python-package/lightgbm/dask.py (hunk 2 above)
Comment (jameslamb) [reply python-package/lightgbm/dask.py:30]: I saw this conversation was marked resolved, but you did not implement what recommended.

Can you please change this back to `int`, and remove `out_port = None `on line 54, and see if it resolves the warning? If you disagree with that, can you please explain why?

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks very much! I left another round of suggestions.

**Comment by jameslamb** [comment]
@tara-jawahar sorry for the inconvenience, but can you please merge the latest changes on the `master` branch into this? I just merged a fairly large change to the Dask module: https://github.com/microsoft/LightGBM/pull/3883

**Comment by tara-jawahar** [comment]
@jameslamb I selectively ignored some mypy errors based on your previous comments

**Comment by jameslamb** [comment]
Thanks very much! Can you please merge the latest changes from the `master` branch?

**Comment by tara-jawahar** [comment]
Hi James, sorry I missed your comment about that before! Thanks for pointing it out. Does it look like I did it correctly?

**Comment by jameslamb** [comment]
> Hi James, sorry I missed your comment about that before! Thanks for pointing it out. Does it look like I did it correctly?

no problem! The merging does not look quite correct. We shouldn't see all these other commits from previous PRs.

I'd be happy to fix it if you'd like. Can I push to your branch?

**Comment by tara-jawahar** [comment]
Yes, go ahead!

**Comment by jameslamb** [comment]
Thanks! I just fixed this. In case you're curious, these are the commands I ran

```shell
git clone git@github.com:tara-jawahar/LightGBM.git tara-lgb
cd tara-lgb
git remote add upstream git@github.com:microsoft/LightGBM.git
git pull upstream master
git fetch origin mypy-errors
git checkout mypy-errors
git merge master
git push origin mypy-errors
```

File: python-package/lightgbm/callback.py [hunk 6]
```diff
@@ -25,8 +25,8 @@ def __init__(self, best_iteration, best_score):
 
 
 # Callback environment used by callbacks
-CallbackEnv = collections.namedtuple(
-    "LightGBMCallbackEnv",
+CallbackEnv = collections.namedtuple( # type: ignore
+    "LightGBMCallbackEnv",
```
Comment (jameslamb) [inline python-package/lightgbm/callback.py:29]:
```suggestion
    "LightGBMCallbackEnv",
```

File: python-package/lightgbm/sklearn.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jameslamb) [reply python-package/lightgbm/sklearn.py:832]:
```suggestion
                   + _base_doc[_base_doc.find('eval_metric :'):])
```

File: python-package/lightgbm/sklearn.py [hunk 7]
```diff
@@ -824,12 +824,12 @@ def fit(self, X, y,
         return self
 
     _base_doc = LGBMModel.fit.__doc__
-    _base_doc = (_base_doc[:_base_doc.find('group :')]
-                 + _base_doc[_base_doc.find('eval_set :'):])
+    _base_doc = (_base_doc[:_base_doc.find('group :')] # type: ignore
+                + _base_doc[_base_doc.find('eval_set :'):]) # type: ignore
```
Comment (jameslamb) [inline python-package/lightgbm/sklearn.py:828]:
```suggestion
                 + _base_doc[_base_doc.find('eval_set :'):]) # type: ignore
```

File: python-package/lightgbm/sklearn.py [hunk 8]
```diff
@@ -824,12 +824,12 @@ def fit(self, X, y,
...
     _base_doc = (_base_doc[:_base_doc.find('eval_class_weight :')]
-                 + _base_doc[_base_doc.find('eval_init_score :'):])
+                + _base_doc[_base_doc.find('eval_init_score :'):])
```
Comment (jameslamb) [inline python-package/lightgbm/sklearn.py:830]:
```suggestion
                 + _base_doc[_base_doc.find('eval_init_score :'):])
```

File: python-package/lightgbm/sklearn.py [hunk 9]
```diff
@@ -897,8 +897,8 @@ def fit(self, X, y,
         return self
 
     _base_doc = LGBMModel.fit.__doc__
-    _base_doc = (_base_doc[:_base_doc.find('group :')]
-                 + _base_doc[_base_doc.find('eval_set :'):])
+    _base_doc = (_base_doc[:_base_doc.find('group :')] # type: ignore
+                 + _base_doc[_base_doc.find('eval_set :'):]) # type: ignore
```
Comment (jameslamb) [inline python-package/lightgbm/sklearn.py:901]: what issue did these `# type: ignore` comments on `_base_doc` fix? I ran `mypy --ignore-missing-imports python-package/` on current `master`, and I don't see any notes related to these lines 

```text
python-package/lightgbm/dask.py:97: error: Incompatible return value type (got "Optional[int]", expected "int")
python-package/lightgbm/dask.py:120: error: Need type annotation for 'lightgbm_ports' (hint: "lightgbm_ports: Set[<type>] = ...")
python-package/lightgbm/dask.py:309: error: "Dict[str, Any]" has no attribute "status"
python-package/lightgbm/dask.py:310: error: Incompatible return value type (got "Dict[str, Any]", expected "LGBMModel")
python-package/lightgbm/dask.py:313: error: "Dict[str, Any]" has no attribute "key"; maybe "keys"?
python-package/lightgbm/dask.py:478: error: "_DaskLGBMModel" has no attribute "_other_params"
python-package/lightgbm/dask.py:496: error: "_DaskLGBMModel" has no attribute "get_params"
python-package/lightgbm/dask.py:510: error: "_DaskLGBMModel" has no attribute "set_params"
python-package/lightgbm/dask.py:516: error: "_DaskLGBMModel" has no attribute "get_params"
python-package/lightgbm/dask.py:525: error: Item "_DaskLGBMModel" of "Union[_DaskLGBMModel, LGBMModel]" has no attribute "get_params"
python-package/lightgbm/dask.py:587: error: Item "None" of "Optional[str]" has no attribute "partition"
python-package/lightgbm/dask.py:602: error: Signature of "fit" incompatible with supertype "LGBMClassifier"
python-package/lightgbm/dask.py:602: error: Signature of "fit" incompatible with supertype "LGBMModel"
python-package/lightgbm/dask.py:610: error: Incompatible return value type (got "_DaskLGBMModel", expected "DaskLGBMClassifier")
python-package/lightgbm/dask.py:636: error: Signature of "predict" incompatible with supertype "LGBMClassifier"
python-package/lightgbm/dask.py:636: error: Signature of "predict" incompatible with supertype "LGBMModel"
python-package/lightgbm/dask.py:654: error: Signature of "predict_proba" incompatible with supertype "LGBMClassifier"
python-package/lightgbm/dask.py:738: error: Item "None" of "Optional[str]" has no attribute "partition"
python-package/lightgbm/dask.py:753: error: Signature of "fit" incompatible with supertype "LGBMRegressor"
python-package/lightgbm/dask.py:753: error: Signature of "fit" incompatible with supertype "LGBMModel"
python-package/lightgbm/dask.py:761: error: Incompatible return value type (got "_DaskLGBMModel", expected "DaskLGBMRegressor")
python-package/lightgbm/dask.py:787: error: Signature of "predict" incompatible with supertype "LGBMModel"
python-package/lightgbm/dask.py:870: error: Item "None" of "Optional[str]" has no attribute "partition"
python-package/lightgbm/dask.py:885: error: Signature of "fit" incompatible with supertype "LGBMRanker"
python-package/lightgbm/dask.py:885: error: Signature of "fit" incompatible with supertype "LGBMModel"
python-package/lightgbm/dask.py:898: error: Incompatible return value type (got "_DaskLGBMModel", expected "DaskLGBMRanker")
python-package/lightgbm/dask.py:928: error: Signature of "predict" incompatible with supertype "LGBMModel"
```

**Review by jameslamb — CHANGES_REQUESTED** [review]
This is looking pretty good, thanks for the help! Whenever you have time, could you answer my one new question about the changes you made to `_doc` elements in `dask.py`?

File: python-package/lightgbm/sklearn.py (hunk 9 above)
Comment (tara-jawahar) [reply python-package/lightgbm/sklearn.py:901]: these are the issues I was ignoring:

Screen Shot 2021-02-14 at 5 06 06 PM

Let me know if that's incorrect and I can take it out!

Comment (jameslamb) [reply python-package/lightgbm/sklearn.py:901]: haha oh! I was looking the Dask module but clearly these changes and the warning come from `sklearn.py`. My mistake, sorry. Thanks for the explanation!

File: python-package/lightgbm/callback.py [hunk 10]
```diff
@@ -25,7 +25,7 @@ def __init__(self, best_iteration, best_score):
 
 
 # Callback environment used by callbacks
-CallbackEnv = collections.namedtuple(
+CallbackEnv = collections.namedtuple( # type: ignore
```
Comment (jameslamb) [inline python-package/lightgbm/callback.py:28]: could you remove this? I think this fix in another PR will handle this: https://github.com/microsoft/LightGBM/pull/3916/files#r571475894

File: python-package/lightgbm/dask.py [hunk 11]
```diff
@@ -306,11 +305,11 @@ def _train(
...
```
Comment (jameslamb) [inline python-package/lightgbm/dask.py:308]:
```suggestion
        if part.status == 'error':  # type: ignore
```

There should be two spaces before every inline comment. This is why the `lint` job in our automated tests is failing.

File: python-package/lightgbm/sklearn.py [hunk 12]
```diff
@@ -824,8 +824,8 @@ def fit(self, X, y,
...
```
Comment (jameslamb) [inline python-package/lightgbm/sklearn.py:828]:
```suggestion
    _base_doc = (_base_doc[:_base_doc.find('group :')]  # type: ignore
                 + _base_doc[_base_doc.find('eval_set :'):])  # type: ignore
```

File: python-package/lightgbm/sklearn.py (hunk 9 above)
Comment (jameslamb) [inline python-package/lightgbm/sklearn.py:901]:
```suggestion
    _base_doc = (_base_doc[:_base_doc.find('group :')]  # type: ignore
                 + _base_doc[_base_doc.find('eval_set :'):])  # type: ignore
```

File: python-package/lightgbm/sklearn.py [hunk 13]
```diff
@@ -989,8 +989,8 @@ def fit(self, X, y,
         return self
 
     _base_doc = LGBMModel.fit.__doc__
-    fit.__doc__ = (_base_doc[:_base_doc.find('eval_class_weight :')]
-                   + _base_doc[_base_doc.find('eval_init_score :'):])
+    fit.__doc__ = (_base_doc[:_base_doc.find('eval_class_weight :')] # type: ignore
+                   + _base_doc[_base_doc.find('eval_init_score :'):]) # type: ignore
```
Comment (jameslamb) [inline python-package/lightgbm/sklearn.py:993]:
```suggestion
    fit.__doc__ = (_base_doc[:_base_doc.find('eval_class_weight :')]  # type: ignore
                   + _base_doc[_base_doc.find('eval_init_score :'):])  # type: ignore
```

File: python-package/lightgbm/callback.py (hunk 10 above)
Comment (jameslamb) [reply python-package/lightgbm/callback.py:28]:
```suggestion
CallbackEnv = collections.namedtuple(
```

**Review by jameslamb — CHANGES_REQUESTED** [review]
This is very close, thanks for the patience and explanations! If you accept the newest round of suggestions I left in the browser, I think tests will pass and this can be merged

**Comment by tara-jawahar** [comment]
Done! Thank you so much for the detailed comments - I'll keep all of these in mind.

**Review by jameslamb — APPROVED** [review]
thanks for the help! I hope you'll come back and contribute in the future.

**Pull request merged**